icu4c/source/common/putil.cpp

```diff
@@ -1085,6 +1094,21 @@ uprv_tzname(int n)
 /* This code can be temporarily disabled to test tzname resolution later on. */
 #ifndef DEBUG_TZNAME
     tzid = getenv("TZ");
+#if U_PLATFORM == U_PF_ANDROID
+    static char timezone[PROP_VALUE_MAX];
+    if (tzid == NULL) {
+        void* libc = dlopen("libc.so", RTLD_NOLOAD);
```

Comment:
Why is this a `dlopen` instead of a regular functioncall?

Reply:
This function is defined in NDK.  And newer platforms (aarch64 etc) don't export it in NDK's libc even if Android bionic libc has/exports this function due to ABI compatibility.

More details is https://issuetracker.google.com/issues/37020706 and https://stackoverflow.com/questions/28413530/.

So we have to use dlsym to use it.

Also, Android O+ has new async function as `__system_property_read_callback`, but this is Android O or later only.

Reply:
I think this bug should get fixed. Alternatively, I could file another PR for __system_property_read_callback  for Android O+ devices.

@enh if he can review the change since the best practice may be considering TZ environment variable on Android, but I think the variable is not set normally.

https://android.googlesource.com/platform/bionic/+blame/master/libc/tzcode/bionic.cpp#45

Reply:
i would just call dlsym() with `RTLD_DEFAULT` rather than using dlopen. i'd avoid the `static` too.

but do you actually need any of this? if you're building with a current NDK, the direct code should just work. and if you're not, _that's_ the bug :-) (iirc this was never removed from the _platform_, but was inaccessible in the L headers back when the NDK had per-platform headers. the current unified headers certainly claim this is always available, as does the version script.)

Reply:
> 
> 
> sorry for publishing the comment late. __system_property_get may be removed in the future. Could you branch the implementation in runtime and call __system_property_read_callback for Android O (API level 26+) if the symbol exists ?

I am sorry for too delay. Acutally, `__system_property_get` is existed even if Android 10 (32bit and 64bit).
`uprv_tzname` isn't async function, so I agree that this should be another PR for `__system_property_read_callback`. To use `__system_property_read_callback`, we need add another hack (adding mutex to wait callbackt?) to get tzname.

Reply:
> Acutally, __system_property_get is existed even if Android 10

i know. i'm the person who'd have to approve removing it from Android, and there are no circumstances under which i'd do that :-)

it's not going away.

i didn't understand the point you were trying to make about `__system_property_read_callback`.

Reply:
__system_property_read_callback is async API, it isn't better to get value immediately. If uprv_tzname is async, we can use it simply.

Reply:
> __system_property_read_callback is async API, it isn't better to get value immediately. If uprv_tzname is async, we can use it simply.

no, it's not asynchronous at all. the only purpose of the callback is to make it clearer what the lifetime of the string is --- it's valid until you return from the callback. any other API would have had memory ownership issues. (and we didn't want to _always_ strdup() and make the caller free().)